diff --git a/drivers/power/supply/qcom/smb5-lib-munch.h b/drivers/power/supply/qcom/smb5-lib-munch.h
--- a/drivers/power/supply/qcom/smb5-lib-munch.h
+++ b/drivers/power/supply/qcom/smb5-lib-munch.h
@@ -72,6 +72,7 @@
 #define SW_THERM_REGULATION_VOTER	"SW_THERM_REGULATION_VOTER"
 #define JEITA_ARB_VOTER			"JEITA_ARB_VOTER"
 #define MOISTURE_VOTER			"MOISTURE_VOTER"
+#define BYPASS_VOTER			"BYPASS_VOTER"
 #define HVDCP2_ICL_VOTER		"HVDCP2_ICL_VOTER"
 #define HVDCP2_FCC_VOTER		"HVDCP2_FCC_VOTER"
 #define AICL_THRESHOLD_VOTER		"AICL_THRESHOLD_VOTER"
diff --git a/drivers/power/supply/qcom/smb5-lib-munch.c b/drivers/power/supply/qcom/smb5-lib-munch.c
--- a/drivers/power/supply/qcom/smb5-lib-munch.c
+++ b/drivers/power/supply/qcom/smb5-lib-munch.c
@@ -43,6 +43,8 @@
 	|| typec_mode == POWER_SUPPLY_TYPEC_SOURCE_HIGH)	\
 	&& (!chg->typec_legacy || chg->typec_legacy_use_rp_icl))
 
+static int bypass_charging = 0;
+
 bool off_charge_flag;
 static bool first_boot_flag;
 
@@ -2401,9 +2403,13 @@
 int smblib_get_prop_input_suspend(struct smb_charger *chg,
 				  union power_supply_propval *val)
 {
-	val->intval
-		= (get_client_vote(chg->usb_icl_votable, USER_VOTER) == 0)
-		 || get_client_vote(chg->dc_suspend_votable, USER_VOTER);
+	if ((get_client_vote(chg->chg_disable_votable, BYPASS_VOTER) == 1)) {
+		val->intval = 1;
+	} else if (bypass_charging) {
+		val->intval = 2;
+	} else {
+		val->intval = 0;
+	}
 	return 0;
 }
 
@@ -3169,20 +3175,37 @@
 	int rc;
 
 	/* vote 0mA when suspended */
-	rc = vote(chg->usb_icl_votable, USER_VOTER, (bool)val->intval, 0);
+	rc = vote(chg->usb_icl_votable, USER_VOTER, false, 0);
 	if (rc < 0) {
 		smblib_err(chg, "Couldn't vote to %s USB rc=%d\n",
 			(bool)val->intval ? "suspend" : "resume", rc);
 		return rc;
 	}
 
-	rc = vote(chg->dc_suspend_votable, USER_VOTER, (bool)val->intval, 0);
+	rc = vote(chg->dc_suspend_votable, USER_VOTER, false, 0);
 	if (rc < 0) {
 		smblib_err(chg, "Couldn't vote to %s DC rc=%d\n",
 			(bool)val->intval ? "suspend" : "resume", rc);
 		return rc;
 	}
 
+	if (val->intval == 1) {
+		rc = vote(chg->chg_disable_votable, BYPASS_VOTER, 1, 0);
+		bypass_charging = 0;
+	} else if (val->intval == 2) {
+		rc = vote(chg->chg_disable_votable, BYPASS_VOTER, 0, 0);
+		bypass_charging = 1;
+	} else {
+		rc = vote(chg->chg_disable_votable, BYPASS_VOTER, 0, 0);
+		bypass_charging = 0;
+	}
+
+	if (rc < 0) {
+		smblib_err(chg, "Couldn't vote to %d input_suspend rc=%d\n",
+			val->intval, rc);
+		return rc;
+	}
+
 	power_supply_changed(chg->batt_psy);
 	return rc;
 }
@@ -3490,6 +3513,9 @@
 	int thermal_fcc_ua = 0;
 	int rc;
 
+	if (bypass_charging)
+		chg->system_temp_level = 0;
+
 	if (chg->system_temp_level >= MAX_TEMP_LEVEL)
 		return 0;
 
@@ -3579,6 +3605,9 @@
 	struct smb_charger *chg = container_of(work, struct smb_charger,
 			thermal_setting_work.work);
 
+	if (bypass_charging)
+		chg->pps_thermal_level = 0;
+
 	if (chg->pps_thermal_level > chg->system_temp_level) {
 		if (chg->pps_thermal_level - chg->system_temp_level >= 2)
 			chg->pps_thermal_level -= 2;
@@ -3638,6 +3667,9 @@
 	vote(chg->chg_disable_votable, THERMAL_DAEMON_VOTER, false, 0);
 #endif
 
+	if (bypass_charging)
+		chg->system_temp_level = 0;
+
 	if (chg->pd_active == POWER_SUPPLY_PD_PPS_ACTIVE)
 		schedule_delayed_work(&chg->thermal_setting_work, 3 * HZ);
 	else
